src/builder.rs

```diff
@@ -207,48 +222,78 @@ impl SpendInfo {
 
     fn prepare<R: RngCore>(self, rng: R) -> PreparedSpendInfo {
         PreparedSpendInfo {
-            proof_generation_key: self.proof_generation_key,
+            fvk: self.fvk,
             note: self.note,
             merkle_path: self.merkle_path,
             rcv: ValueCommitTrapdoor::random(rng),
-            dummy_ask: self.dummy_ask,
+            dummy_expsk: self.dummy_expsk,
         }
     }
 }
 
 #[derive(Debug, Clone)]
 struct PreparedSpendInfo {
-    proof_generation_key: ProofGenerationKey,
+    fvk: FullViewingKey,
     note: Note,
     merkle_path: MerklePath,
     rcv: ValueCommitTrapdoor,
-    dummy_ask: Option<SpendAuthorizingKey>,
+    dummy_expsk: Option<ExpandedSpendingKey>,
 }
 
 impl PreparedSpendInfo {
-    fn build<Pr: SpendProver, R: RngCore>(
-        self,
+    fn build_inner<R: RngCore>(
+        &self,
         mut rng: R,
-    ) -> Result<SpendDescription<InProgress<Unproven, Unsigned>>, Error> {
+    ) -> (
+        ValueCommitment,
+        Nullifier,
+        redjubjub::VerificationKey<SpendAuth>,
+        jubjub::Scalar,
+    ) {
         // Construct the value commitment.
         let alpha = jubjub::Fr::random(&mut rng);
         let cv = ValueCommitment::derive(self.note.value(), self.rcv.clone());
-        let node = Node::from_cmu(&self.note.cmu());
-        let anchor = *self.merkle_path.root(node).inner();
 
-        let ak = self.proof_generation_key.ak.clone();
+        let ak = self.fvk.vk.ak.clone();
 
         // This is the result of the re-randomization, we compute it for the caller
         let rk = ak.randomize(&alpha);
 
         let nullifier = self.note.nf(
-            &self.proof_generation_key.to_viewing_key().nk,
+            &self.fvk.vk.nk,
             u64::try_from(self.merkle_path.position())
                 .expect("Sapling note commitment tree position must fit into a u64"),
         );
 
+        (cv, nullifier, rk, alpha)
+    }
+
+    fn build<Pr: SpendProver, R: RngCore>(
+        self,
+        proof_generation_key: Option<ProofGenerationKey>,
+        rng: R,
+    ) -> Result<SpendDescription<InProgress<Unproven, Unsigned>>, Error> {
+        let proof_generation_key = match (proof_generation_key, self.dummy_expsk.as_ref()) {
+            (Some(proof_generation_key), None) => Ok(proof_generation_key),
+            (None, Some(expsk)) => Ok(expsk.proof_generation_key()),
+            (Some(_), Some(_)) => Err(Error::WrongSpendingKey),
```

Comment:
Should the case where `proof_generation_key` is `Some` and equal to `expsk.proof_generation_key()` be permitted?
```suggestion
            (Some(proof_generation_key), Some(expsk)) if proof_generation_key == expsk.proof_generation_key() => Ok(proof_generation_key),
            (Some(_), Some(_)) => Err(Error::WrongSpendingKey),
```

Reply:
`self.dummy_expsk` is generated internally from the RNG, and isn't directly extractable.